plugins/woocommerce/src/Internal/Admin/ProductDownloadsPreview.php

```diff
@@ -0,0 +1,182 @@
+	 *
+	 * @since 9.9.0
+	 */
+	public function register_rest_routes() {
+		$namespace     = 'wc/v3';
+		$route_base    = '/admin/product-downloads-preview';
+		$route_pattern = '/(?P<product_id>[\d]+)/(?P<attachment_id>[\d]+)';
+
+		$args = array(
+			'product_id'    => array(
+				'required'    => true,
+				'type'        => 'integer',
+				'description' => 'Product ID that the downloadable image belongs to',
+			),
+			'attachment_id' => array(
+				'required'    => true,
+				'type'        => 'integer',
+				'description' => 'Attachment ID to preview',
+			),
+			'size'          => array(
+				'type'        => 'string',
+				'default'     => 'large',
+				'description' => 'Image size to display',
+			),
+		);
+
+		register_rest_route(
+			$namespace,
+			$route_base . $route_pattern,
+			array(
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => array( $this, 'get_preview' ),
+				'permission_callback' => array( $this, 'get_preview_permissions_check' ),
+				'args'                => $args,
+			)
+		);
+	}
+
+	/**
+	 * Permission check for REST API endpoint.
+	 *
+	 * @since 9.9.0
+	 * @param \WP_REST_Request $request Request details.
+	 * @return bool|\WP_Error
+	 */
+	public function get_preview_permissions_check( $request ) {
+		if ( ! current_user_can( 'manage_woocommerce' ) ) {
+			return new WP_Error(
+				'woocommerce_rest_unauthorized',
+				__( 'Unauthorized access.', 'woocommerce' ),
+				array( 'status' => 403 )
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Serve the preview image
+	 *
+	 * @since 9.9.0
+	 * @param \WP_REST_Request $request Request details.
+	 * @return \WP_REST_Response|\WP_Error
+	 */
+	public function get_preview( $request ) {
+		$attachment_id  = $request['attachment_id'];
+		$product_id     = $request['product_id'];
+		$requested_size = $request['size'] ?? 'large';
+
+		$file_path = get_attached_file( $attachment_id );
+		if ( ! $file_path || ! is_readable( $file_path ) ) {
+			return new WP_Error(
+				'woocommerce_rest_file_not_found',
+				__( 'File not found', 'woocommerce' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		$mime_type          = get_post_mime_type( $attachment_id );
+		$allowed_mime_types = array(
+			'image/jpeg',
+			'image/jpg',
+			'image/png',
+			'image/gif',
+			'image/webp',
+		);
+
+		if ( ! in_array( $mime_type, $allowed_mime_types, true ) ) {
+			return new WP_Error(
+				'woocommerce_rest_invalid_file_type',
+				__( 'Invalid file type', 'woocommerce' ),
+				array( 'status' => 403 )
+			);
+		}
+
+		$size = 'full' === $requested_size ? 'large' : $requested_size;
+
+		$resized = image_get_intermediate_size( $attachment_id, $size );
+		if ( $resized && isset( $resized['path'] ) ) {
+			$uploads_dir       = wp_upload_dir();
+			$resized_file_path = $uploads_dir['basedir'] . '/' . $resized['path'];
+			if ( is_readable( $resized_file_path ) ) {
+				$file_path = $resized_file_path;
+			}
+		}
+
+		// Clean any existing output.
+		ob_clean();
```

Comment:
This is probably fine 95% of the time, but will only clean a single level of output bufferring. I think there was a more complete version in an earlier iteration of the PR: could it be worth revisiting that?

Reply:
Good point, I've updated it here: 6ef4e7be4f0c94921c7665d47f265ba7862ac091

Reply:
I've simplified it a bit from the previous solution.